Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1015,13 +1018,127 @@ void CGContextClipToRects(CGContextRef context, const CGRect* rects, unsigned co
+
+    newOpacityBrush->SetTransform(__CGAffineTransformToD2D_F(transform));
+
+    // Compliance with reference platform: D2D extends the last pixel of the opacity brush out to the
+    // ends of the earth. The reference platform truncates the clipping region at the edge of the mask.
+    ComPtr<ID2D1RectangleGeometry> rectGeometry;
+    ComPtr<ID2D1TransformedGeometry> transformedRectClippingGeometry;
+    RETURN_IF_FAILED(Factory()->CreateRectangleGeometry(__CGRectToD2D_F(rect), &rectGeometry));
+    RETURN_IF_FAILED(Factory()->CreateTransformedGeometry(rectGeometry.Get(),
+                                                          __CGAffineTransformToD2D_F(userToDeviceTransform),
+                                                          &transformedRectClippingGeometry));
+
+    if (state.opacityBrush) {
+        // If we already have an opacity brush, we have to go to great lengths to compose the two clipping images.
+        // - Create a compatible render target with a backing bitmap
+        // - Using two layers (reasons detailed below), fill a geometry with the intersection of the two 
+        //   masks.
+        // - Use that bitmap (untransformed, as the render has resolved the global coordinate system conflict)
+        //   as the opacity brush for future drawing.
+        ComPtr<ID2D1BitmapRenderTarget> compatibleTarget;
+        RETURN_IF_FAILED(deviceContext->CreateCompatibleRenderTarget(&compatibleTarget));
+        ComPtr<ID2D1DeviceContext> compatibleContext;
+        RETURN_IF_FAILED(compatibleTarget.As(&compatibleContext));
+
+        compatibleContext->BeginDraw();
+        compatibleContext->PushLayer(D2D1::LayerParameters(D2D1::InfiniteRect(),
+                                                       nullptr,
+                                                       D2D1_ANTIALIAS_MODE_PER_PRIMITIVE,
+                                                       D2D1::IdentityMatrix(),
+                                                       1.0, // 1.0 global alpha for brush composition
+                                                       state.opacityBrush.Get()),
+                                 nullptr);
+        compatibleContext->PushLayer(D2D1::LayerParameters(D2D1::InfiniteRect(),
+                                                       transformedRectClippingGeometry.Get(),
+                                                       D2D1_ANTIALIAS_MODE_PER_PRIMITIVE,
+                                                       D2D1::IdentityMatrix(),
+                                                       1.0, // 1.0 global alpha for brush composition
+                                                       newOpacityBrush.Get()),
+                                 nullptr);
+
+        ComPtr<ID2D1SolidColorBrush> brush;
+        RETURN_IF_FAILED(compatibleContext->CreateSolidColorBrush({ 0, 0, 0, 1 }, &brush));
+        // FillGeometry takes three parameters:
```

Comment:
>FillGeometry [](start = 11, length = 12)

nit, This comment sounds great for the design doc/follow up questions to D2D but seems a bit excessive for in code comments. You could shorten it to something like "Using two stacked layers as brush opacity results in failure of other, preferable APIs."